dawn/src/dawn/CodeGen/F90Util.h

```diff
@@ -0,0 +1,177 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#pragma once
+
+#include <algorithm>
+#include <functional>
+#include <optional>
+#include <sstream>
+#include <vector>
+#define DAWN_FORTRAN_INDENT 2
+
+namespace dawn {
+namespace codegen {
+
+namespace {
+void indent(int level, std::stringstream& ss) {
+  ss << std::string(DAWN_FORTRAN_INDENT * level, ' ');
+}
+struct endlineType {};
+static endlineType endline;
+} // namespace
+
+class IndentedStringStream {
+protected:
+  std::reference_wrapper<std::stringstream> ss_;
+  int indentLevel_ = 0;
+  bool freshNewLine_ = true;
+
+public:
+  IndentedStringStream(std::stringstream& s, int il = 0) : ss_(s), indentLevel_(il) {}
+
+  IndentedStringStream& operator<<(endlineType) {
+    ss() << std::endl;
+    freshNewLine_ = true;
+    return *this;
+  }
+
+  template <class T>
+  IndentedStringStream& operator<<(T&& data) {
+    if(freshNewLine_)
+      indent(indentLevel_, ss());
+    freshNewLine_ = false;
+    ss() << data;
+    return *this;
+  }
+
+  void increaseIndent() { indentLevel_++; }
+  void decreaseIndent() {
+    if(indentLevel_ > 0)
+      indentLevel_--;
+  }
+  void setIndentLevel(int il) { indentLevel_ = il; }
+
+  std::stringstream& ss() { return ss_.get(); }
+
+  std::string str() { return ss().str(); }
+};
+
+class FortranInterfaceAPI {
+public:
+  enum class InterfaceType { INTEGER, FLOAT, DOUBLE, CHAR, OBJ };
+  FortranInterfaceAPI(std::string name, std::optional<InterfaceType> returnType = std::nullopt)
```

Comment:
I don't understand why you need the option with no return type. I don't think you use it, and I don't quite get what would be generated if the return type is `nullopt`

Reply:
You are right, there's currently no use case. An API that wouldn't need a return type is a setter for globals
```
set_globalname(double value) { ... }
```
that would need to be generated into a `subroutine` and not a `function` (Fortran is very strange to me in this matter).
Do we keep it, just in case?